FOR SH_SHARE, LOCK_DATA_PSL is raising: Feature not enabled (pycurl-7.47.0)

Hi,

Since i have switched to newer version 7.47, my build pipelines started showing error:

self._share_handle.setopt(pycurl.SH_SHARE, pycurl.LOCK_DATA_PSL)
E       pycurl.error: (5, 'Feature not enabled in this library')

Package: pycurl-7.47.0-cp313-cp313-manylinux_2_28_x86_64.whl


Unfortunately, I am unable to find the root cause for this. Any help would be much appreciated.

--- a/.github/workflows/ci-macos.yml
+++ b/.github/workflows/ci-macos.yml
@@ -25,7 +25,7 @@ jobs:
     env:
       PYCURL_CURL_CONFIG: /usr/bin/curl-config
       PYCURL_SSL_LIBRARY: sectransp
-      VCPKG_BINARY_SOURCES: 'clear;nuget,GitHub,readwrite'
+      VCPKG_BINARY_SOURCES: "clear;nuget,GitHub,${{ github.event_name == 'pull_request' && 'read' || 'readwrite' }}"
     strategy:
       fail-fast: false
       matrix:

--- a/.github/workflows/ci-windows.yml
+++ b/.github/workflows/ci-windows.yml
@@ -63,7 +63,7 @@ jobs:
           setapikey "${{ secrets.GITHUB_TOKEN }}" \
           -source "https://nuget.pkg.github.com/${GITHUB_REPOSITORY_OWNER}/index.json"
     - name: Install packages
-      run: vcpkg install curl[brotli,core,http2,non-http,ssh,ssl,websockets]:${{ matrix.arch }}-windows
+      run: vcpkg install curl[brotli,core,http2,non-http,psl,ssh,ssl,websockets]:${{ matrix.arch }}-windows
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip

--- a/.github/workflows/ci-windows.yml
+++ b/.github/workflows/ci-windows.yml
@@ -23,7 +23,7 @@ jobs:
 
     runs-on: ${{ matrix.os }}
     env:
-      VCPKG_BINARY_SOURCES: 'clear;nuget,GitHub,readwrite'
+      VCPKG_BINARY_SOURCES: "clear;nuget,GitHub,${{ github.event_name == 'pull_request' && 'read' || 'readwrite' }}"
     strategy:
       matrix:
         os: ["windows-2025"]

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -23,7 +23,7 @@ jobs:
 
     runs-on: ubuntu-24.04
     env:
-      VCPKG_BINARY_SOURCES: 'clear;nuget,GitHub,readwrite'
+      VCPKG_BINARY_SOURCES: "clear;nuget,GitHub,${{ github.event_name == 'pull_request' && 'read' || 'readwrite' }}"
     strategy:
       fail-fast: false
       matrix:

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -67,7 +67,7 @@ jobs:
 
       - name: Install packages (Windows)
         if: runner.os == 'Windows'
-        run: vcpkg install curl[brotli,core,http2,non-http,ssh,ssl,websockets]:${{ matrix.arch }}-windows
+        run: vcpkg install curl[brotli,core,http2,non-http,psl,ssh,ssl,websockets]:${{ matrix.arch }}-windows
 
       - name: Install cibuildwheel
         run: python -m pip install cibuildwheel

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -77,7 +77,7 @@ before-all = [
     "git clone https://github.com/microsoft/vcpkg.git $VCPKG_ROOT",
     "$VCPKG_ROOT/bootstrap-vcpkg.sh",
     "{project}/scripts/setup-vcpkg-nuget.sh",
-    "PATH=$(echo \"$PATH\" | tr ':' '\n' | grep -v '^/usr/local/bin$' | paste -sd:) vcpkg install curl[brotli,core,gssapi,http2,non-http,openssl,ssh,websockets] --triplet=$VCPKG_TRIPLET",
+    "PATH=$(echo \"$PATH\" | tr ':' '\n' | grep -v '^/usr/local/bin$' | paste -sd:) vcpkg install curl[brotli,core,gssapi,http2,non-http,openssl,psl,ssh,websockets] --triplet=$VCPKG_TRIPLET",
 ]
 before-test = [
     "pip install flake8 -r requirements-dev.txt",
@@ -104,7 +104,8 @@ environment.VCPKG_TRIPLET = "arm64-linux-dynamic"
 before-all = [
     "if ! which mono >/dev/null 2>&1; then brew install mono; fi",
     "if ! which vcpkg >/dev/null 2>&1; then brew install vcpkg && git clone https://github.com/microsoft/vcpkg.git $HOME/vcpkg; fi",
-    "vcpkg install curl[brotli,core,http2,non-http,openssl,ssh,websockets] --triplet=$VCPKG_TRIPLET",
+    "brew install autoconf autoconf-archive automake libtool",
+    "vcpkg install curl[brotli,core,http2,non-http,openssl,psl,ssh,websockets] --triplet=$VCPKG_TRIPLET",
 ]
 before-test = [
     "pip install flake8 -r requirements-dev.txt",

--- a/src/module.c
+++ b/src/module.c
@@ -1773,6 +1773,9 @@ PYCURL_IGNORE_DEPRECATED_END
 #if LIBCURL_VERSION_NUM >= MAKE_LIBCURL_VERSION(7, 61, 0)
     insint_s(d, "LOCK_DATA_PSL", CURL_LOCK_DATA_PSL);
 #endif
+#if LIBCURL_VERSION_NUM >= MAKE_LIBCURL_VERSION(7, 88, 0)
+    insint_s(d, "LOCK_DATA_HSTS", CURL_LOCK_DATA_HSTS);
+#endif
 
     /* Initialize callback locks if ssl is enabled */
 #if defined(PYCURL_NEED_SSL_TSL)

--- a/src/share.c
+++ b/src/share.c
@@ -125,6 +125,9 @@ share_setopt_one(CurlShareObject *self, int option, long data_kind)
 #if LIBCURL_VERSION_NUM >= MAKE_LIBCURL_VERSION(7, 61, 0)
     case CURL_LOCK_DATA_PSL:
 #endif
+#if LIBCURL_VERSION_NUM >= MAKE_LIBCURL_VERSION(7, 88, 0)
+    case CURL_LOCK_DATA_HSTS:
+#endif
         break;
     default:
         PyErr_SetString(PyExc_TypeError, "invalid arguments to setopt");

--- a/tests/share_test.py
+++ b/tests/share_test.py
@@ -343,3 +343,22 @@ def test_concurrent_api_calls_do_not_crash():
             f.result()
 
     s.close()
+
+
+@util.min_libcurl(7, 57, 0)
+def test_share_setopt_lock_data_connect():
+    s = pycurl.CurlShare()
+    s.setopt(pycurl.SH_SHARE, pycurl.LOCK_DATA_CONNECT)
+
+
+@util.min_libcurl(7, 61, 0)
+@util.only_psl
+def test_share_setopt_lock_data_psl():
+    s = pycurl.CurlShare()
+    s.setopt(pycurl.SH_SHARE, pycurl.LOCK_DATA_PSL)
+
+
+@util.min_libcurl(7, 88, 0)
+def test_share_setopt_lock_data_hsts():
+    s = pycurl.CurlShare()
+    s.setopt(pycurl.SH_SHARE, pycurl.LOCK_DATA_HSTS)

--- a/tests/util.py
+++ b/tests/util.py
@@ -291,6 +291,19 @@ def only_tls_srp(fn):
     return decorated
 
 
+def only_psl(fn):
+    import pycurl
+
+    @functools.wraps(fn)
+    def decorated(*args, **kwargs):
+        if not pycurl.version_info()[4] & pycurl.VERSION_PSL:
+            raise unittest.SkipTest("libcurl does not support libpsl")
+
+        return fn(*args, **kwargs)
+
+    return decorated
+
+
 def guard_unknown_libcurl_option(fn):
     """Converts curl error 48, CURLE_UNKNOWN_OPTION, into a SkipTest
     exception. This is meant to be used with tests exercising libcurl

